Source/Particles/Resampling/VelocityCoincidenceThinning.cpp

```diff
@@ -79,21 +98,60 @@ void VelocityCoincidenceThinning::operator() (WarpXParIter& pti, const int lev,
     );
 
     // create a GPU vector to hold the momentum cluster index for each particle
-    amrex::Gpu::DeviceVector<int> momentum_bin_number(bins.numItems());
+    amrex::Gpu::DeviceVector<int> momentum_bin_number(n_parts_in_tile);
     auto* momentum_bin_number_data = momentum_bin_number.data();
 
     // create a GPU vector to hold the index sorting for the momentum bins
-    amrex::Gpu::DeviceVector<int> sorted_indices(bins.numItems());
+    amrex::Gpu::DeviceVector<int> sorted_indices(n_parts_in_tile);
     auto* sorted_indices_data = sorted_indices.data();
 
-    const auto Ntheta = m_ntheta;
-    const auto Nphi = m_nphi;
-
-    const auto dr = m_delta_ur;
-    const auto dtheta = 2.0_prt * MathConst::pi / Ntheta;
-    const auto dphi = MathConst::pi / Nphi;
     constexpr auto c2 = PhysConst::c * PhysConst::c;
 
+    auto velocityBinCalculator = VelocityBinCalculator();
+    velocityBinCalculator.velocity_grid_type = m_velocity_grid_type;
+    if (m_velocity_grid_type == VelocityGridType::Spherical) {
+        velocityBinCalculator.dur = m_delta_ur;
+        velocityBinCalculator.n1 = m_ntheta;
+        velocityBinCalculator.n2 = m_nphi;
+        velocityBinCalculator.dutheta = 2.0_prt * MathConst::pi / m_ntheta;
+        velocityBinCalculator.duphi = MathConst::pi / m_nphi;
+    }
+    else if (m_velocity_grid_type == VelocityGridType::Cartesian) {
+        velocityBinCalculator.dux = m_delta_ux;
+        velocityBinCalculator.duy = m_delta_uy;
+        velocityBinCalculator.duz = m_delta_uz;
+
+        // get the minimum and maximum velocities to determine the velocity space
+        // grid boundaries
+        using PType = typename WarpXParticleContainer::SuperParticleType;
+        velocityBinCalculator.ux_min = ReduceMin( *pc,
+            [=] AMREX_GPU_HOST_DEVICE (const PType& p)
+            { return p.rdata(PIdx::ux); }
+        );
+        velocityBinCalculator.uy_min = ReduceMin( *pc,
+            [=] AMREX_GPU_HOST_DEVICE (const PType& p)
+            { return p.rdata(PIdx::uy); }
+        );
+        velocityBinCalculator.uz_min = ReduceMin( *pc,
+            [=] AMREX_GPU_HOST_DEVICE (const PType& p)
+            { return p.rdata(PIdx::uz); }
+        );
+        velocityBinCalculator.ux_max = ReduceMax( *pc,
+            [=] AMREX_GPU_HOST_DEVICE (const PType& p)
+            { return p.rdata(PIdx::ux); }
+        );
+        velocityBinCalculator.uy_max = ReduceMax( *pc,
+            [=] AMREX_GPU_HOST_DEVICE (const PType& p)
+            { return p.rdata(PIdx::uy); }
+        );
```

Comment:
@WeiqunZhang this is the section in which I would like to perform a reduction over the particle tile rather than the whole particle container.

Reply:
Something like
```
const int np = pti.numParticles();
GpuArray<ParticleReal, AMREX_SPACEDIM*2> result;
{
    using ReduceOpsT = TypeMultiplier<ReduceOps,
                                      ReduceMin[AMREX_SPACEDIM],
                                      ReduceMax[AMREX_SPACEDIM]>;
    using ReduceDataT = TypeMultiplier<ReduceData, ParticleReal[AMREX_SPACEDIM*2]>;
    ReduceOpsT reduce_op;
    ReduceDataT reduce_data(reduce_op);
    using ReduceTuple = typename ReduceDataT::Type;
    reduce_op.eval(n, reduce_data, [=] AMREX_GPU_DEVICE(int i) -> ReduceTuple {
        ....
        return {AMREX_D_DECL(ux,uy,uz),
                AMREX_D_DECL(ux,uy,uz)};
    });
    auto hv = reduce_data.value(reduce_op);
    result = tupleToArray(hv);
}
```

Reply:
Cool! Thanks so much for your help!

Reply:
We probably should add a function to amrex to make this relatively easier for the users. Something like
```
auto result = amrex::ParReduce(TypeList<ReduceOpMin[AMREX_SPACEDIM],ReduceOpMax[AMREX_SPACEDIM]>{},
                                  TypeList<ParticleReal[AMERX_SPACEDIM*2]>{},
                                  np, [=] AMREX_GPU_DEVICE (int i) -> TypeMultiplier<GpuTuple, ParticleReal[AMREX_SPACEDIM*2]> { return ...; });
```

Reply:
Such a simplified function would be helpful. For example, it could improve the code in https://github.com/ECP-WarpX/WarpX/blob/development/Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp. Note however, that in both that case and my present case we need the min and max for each direction rather than the overall min and max (and I actually don't need the max of `uz`). I modified the code you suggested accordingly.
Thanks again for your help!